include/gsl/span

```diff
@@ -121,139 +122,165 @@ namespace details
+        {
+            return rhs - n;
+        }
 
-        constexpr difference_type operator-(span_iterator rhs) const noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr difference_type operator-(const span_iterator<type2>& rhs) const noexcept
         {
-            Expects(span_ == rhs.span_);
-            return index_ - rhs.index_;
+            Expects(begin_ == rhs.begin_);
+            return current_ - rhs.current_;
         }
 
-        constexpr reference operator[](difference_type n) const noexcept { return *(*this + n); }
+        constexpr reference operator[](const difference_type n) const noexcept
+        {
+            return *(*this + n);
+        }
 
-        constexpr friend bool operator==(span_iterator lhs, span_iterator rhs) noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr bool operator==(const span_iterator<type2>& rhs) const noexcept
         {
-            return lhs.span_ == rhs.span_ && lhs.index_ == rhs.index_;
+            return begin_ == rhs.begin_ && current_ == rhs.current_;
         }
 
-        constexpr friend bool operator!=(span_iterator lhs, span_iterator rhs) noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr bool operator!=(const span_iterator<type2>& rhs) const noexcept
         {
-            return !(lhs == rhs);
+            return !(*this == rhs);
         }
 
-        constexpr friend bool operator<(span_iterator lhs, span_iterator rhs) noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr bool operator<(const span_iterator<type2>& rhs) const noexcept
         {
-            return lhs.index_ < rhs.index_;
+            Expects(begin_ == rhs.begin_);
+            return current_ < rhs.current_;
         }
 
-        constexpr friend bool operator<=(span_iterator lhs, span_iterator rhs) noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr bool operator>(const span_iterator<type2>& rhs) const noexcept
         {
-            return !(rhs < lhs);
+            return !(*this < rhs);
         }
 
-        constexpr friend bool operator>(span_iterator lhs, span_iterator rhs) noexcept
+        template <
+            class type2,
+            std::enable_if_t<std::is_same<std::remove_cv_t<type2>, value_type>::value, int> = 0>
+        constexpr bool operator<=(const span_iterator<type2>& rhs) const noexcept
         {
-            return rhs < lhs;
+            return *this < rhs || *this == rhs;
```

Comment:
This is an inefficient way to express the operation (due to repeated validation). It is better to transform this into a single call to `operator<`. `left <= right` is `!(right < left)`.